### PR TITLE
fv3fit: add dense model using PureKerasModel

### PR DESCRIPTION
**Pull request by AnnaKwa**
This PR adds a training function for a dense model that uses the same `DenseHyperparameters` as the existing `DenseModel` but returns a predictor that is a `PureKerasModel` instead of the old `DenseModel`. This will allow single column dense models to be built in a similar manner and utilize the same normalization and clipping features as newer keras models, which will make future development and model training easier. 

Added public API:
- training function for type `dense` 

Refactored public API:
- I changed the `loss` field of the existing `DenseHyperparameters` used by the older `DenseModel` and this newer `dense` model to expect a `LossConfig`. The older `DenseModel` requires `scaling=standard` to work in accordance with its existing behavior. I adjusted `DenseModel.load` to be backwards compatible with older models that have a string entry for their loss.
- Removed the older `DenseModel` training function (not the class, just the training function).

Significant internal changes:
- moved `get_stacked_metadata` to `keras._models.shared.utils` since both convolutional and dense model use it
- deleted an unused `_fill_defaults` function in `keras._models.models`

- [x] Tests added

I tested training a model with the same hyperparameters for the new `dense` training function vs the older `DenseModel`, the results are similar but not identical:
image

### CODE REVIEW

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+            activation="linear",
+            name=f"dense_network_output_{i}",
+        )(hidden_outputs[-1])
+        for i, array in enumerate(y)
+    ]
+    if config.nonnegative_outputs is True:
+        norm_output_layers = [
+            tf.keras.layers.Activation(tf.keras.activations.relu)(output_layer)
+            for output_layer in norm_output_layers
+        ]
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/dense.py:216]: This is a bug - you need to apply this activation to `denorm_output_layers`. Non-negativity in normalized space doesn't guarantee non-negativity in denormalized space.

**Review by mcgibbon — COMMENTED** [review]
Looking great! One bug, one change and some suggestions.

File: external/fv3fit/fv3fit/keras/_models/dense.py (hunk 1 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py:216]: Thanks, good catch

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+            activation=tf.keras.activations.relu,
+            kernel_regularizer=config.dense_network.kernel_regularizer.instance,
+            name=f"hidden_{i}",
+        )
+        if config.dense_network.spectral_normalization:
+            hidden_layer = tfa.layers.SpectralNormalization(
+                hidden_layer, name=f"spectral_norm_{i}"
+            )
+        x = hidden_layer(x)
+        hidden_outputs.append(x)
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py:202]: Oh cool, thanks.

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    ]
+    norm_input_layers = standard_normalize(
+        names=config.input_variables,
+        layers=input_layers,
+        arrays=[_ensure_2d(array) for array in X],
+    )
+    if len(norm_input_layers) > 1:
+        full_input = tf.keras.layers.Concatenate()(norm_input_layers)
+    else:
+        full_input = norm_input_layers[0]
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py:181]: Yes, I think if the helper functions that take in `X/y` for use in getting layer shapes take in the `X_2d` etc. that already have the feature dimension standardized, we can use the same generic helpers across training functions. Then each training function just needs its specifc `ensure_some_dimensionality` function in its module.

File: external/fv3fit/fv3fit/keras/_models/models.py [hunk 4]
```diff
@@ -182,7 +124,13 @@ def __init__(
             self.weights = hyperparameters.weights
         self._normalize_loss = hyperparameters.normalize_loss
         self._optimizer = hyperparameters.optimizer_config.instance
-        self._loss = hyperparameters.loss
+
+        if hyperparameters.loss.scaling != "standard":
+            raise ValueError(
+                "Only 'standard' loss scaling is supported for DenseModel."
+            )
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/models.py:113]: Thanks, I deleted the old training function and replaced it with `dense` in the end to end tests.

File: external/fv3fit/fv3fit/keras/_models/dense.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py]: The denormalization bits were model-specific enough that I left them as is. I put the standardization and concatenation of input layers into a helper function `full_standard_normalized_input` (commit 0c92746984f79b7fe86e97a406cd8048ac69a317)

File: external/fv3fit/fv3fit/keras/_models/convolutional.py [hunk 5]
```diff
@@ -125,19 +111,8 @@ def train_convolutional_model(
...
-    """
-    if len(arr.shape) == 3:
-        return 1
-    else:
-        return arr.shape[3]
-
-
 def _ensure_4d(array: np.ndarray) -> np.ndarray:
-    # ensures [sample, x, y, z] dimensionality
+    # Ensures [sample, x, y, z] dimensionality.
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/convolutional.py:115]: You don't need to change it back, but I prefer the other style for short declarations. Even with punctuation, this isn't a complete sentence because it has no subject.

File: external/fv3fit/fv3fit/keras/_models/convolutional.py [hunk 6]
```diff
@@ -171,38 +132,31 @@ def build_model(
         X: example input for keras fitting, used to determine shape and normalization
         y: example output for keras fitting, used to determine shape and normalization
     """
-    input_layer_shapes = _get_input_layer_shapes(X)
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/convolutional.py:174]: Really liking the updates here.

**Review by mcgibbon — APPROVED** [review]
Great update! You should add some comments as code or docstring comments to the DenseModel talking about how it's only still here for loading old models and not for training new ones. After that, looks good to merge.
